init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
There are a few things I am not entirely happy about:
1) requirement of the warmup iterations per se
2) state machine resets by (newly required) call to cudnn.reset(). I see no way around the need for explicit initialization on cudnn level - I think I should rename it to something more specific or use cudnn.find.reset()
3) state machine now deduces iterations based on repeating calls to the same layer. I find it fragile. It would be better if it would reach to some global net state that counts iterations (reset() could be tied up there as well). Any ideas?
4) Failures to allocate desired workspace should be handled more gradually. I have an upcoming cutorch PR to help with that
5) Workspace is now being kept all the time - may not work very well  on shared systems, but reallocation of a giant slab on every convolution is both expensive and risky. I am adding CNMEM option to cutorch that may be able to help with that, too.